source/base/mpi.cc

```diff
@@ -812,7 +814,890 @@ namespace Utilities
+          AssertThrowMPI(ierr);
+#else
+          (void)owned_indices;
+          (void)comm;
+#endif
+        }
+
+        unsigned int
+        dof_to_dict_rank(const types::global_dof_index i)
+        {
+          return i / dofs_per_process;
+        }
+
+      private:
+        void
+        partition(const IndexSet &owned_indices, const MPI_Comm &comm)
+        {
+#ifdef DEAL_II_WITH_MPI
+          unsigned int n_procs = n_mpi_processes(comm);
+          unsigned int my_rank = this_mpi_process(comm);
+
+          size               = owned_indices.size();
+          dofs_per_process   = (size + n_procs - 1) / n_procs;
+          local_range.first  = std::min(dofs_per_process * my_rank, size);
+          local_range.second = std::min(dofs_per_process * (my_rank + 1), size);
+          local_size         = local_range.second - local_range.first;
+
+          actually_owning_ranks.resize(local_size);
+#else
+          (void)owned_indices;
+          (void)comm;
+#endif
+        }
+      };
+
+      class ConsensusAlgorithmProcess
+        : public dealii::Utilities::MPI::
+            ConsensusAlgorithmProcess<types::global_dof_index, unsigned int>
+      {
+      public:
+        ConsensusAlgorithmProcess(const IndexSet &           owned_indices,
+                                  const IndexSet &           indices_to_look_up,
+                                  const MPI_Comm &           comm,
+                                  std::vector<unsigned int> &array)
+          : owned_indices(owned_indices)
+          , indices_to_look_up(indices_to_look_up)
+          , comm(comm)
+          , array(array)
+        {
+          my_rank = this_mpi_process(comm);
+          n_procs = n_mpi_processes(comm);
+
+          this->dict.reinit(owned_indices, comm);
+        }
+
+        ~ConsensusAlgorithmProcess() = default;
+
+        const IndexSet &           owned_indices;
+        const IndexSet &           indices_to_look_up;
+        const MPI_Comm &           comm;
+        std::vector<unsigned int> &array;
+
+        Dictionary   dict;
+        unsigned int my_rank;
+        unsigned int n_procs;
+
+        std::map<unsigned int, std::vector<types::global_dof_index>> temp;
+        std::map<unsigned int, std::vector<unsigned int>> recv_indices;
+
+        void
+        process_request(const unsigned int                          other_rank,
+                        const std::vector<types::global_dof_index> &buffer_recv,
+                        std::vector<unsigned int> &request_buffer) override
+        {
+          (void)other_rank;
+          Assert(buffer_recv.size() % 2 == 0, ExcInternalError());
+          for (unsigned int j = 0; j < buffer_recv.size(); j += 2)
+            for (auto i = buffer_recv[j]; i < buffer_recv[j + 1]; i++)
+              request_buffer.push_back(
+                dict.actually_owning_ranks[i - dict.local_range.first]);
+        }
+
+        std::vector<unsigned int>
+        compute_targets() override
```

Comment:
```suggestion
        virtual compute_targets() override
```
We mark all overriden (redundantly) as `virtual`.